### ISSUE TITLE
Cannot extract clipping path from JPEG with multiple APP13 segments

### ISSUE THREAD

**Issue by aerpenbeck** [haraldk/TwelveMonkeys#588]
Cannot extract clipping path from a JPEG file with multiple APP13/Photoshop 3.0 segments.

Here's an [example](https://drive.google.com/file/d/1Z1vA7fkA9WTXWRG1vo7mrp6tnQDbMSSm/view?usp=sharing) with 3 APP13 segments.

Using `exiftool` I can verify that the file actually does contain clipping path information.

It's probably related to this part of `com.twelvemonkeys.imageio.path.Paths#readPath` where only the first segment is evaluated:
```
List<JPEGSegment> photoshop = JPEGSegmentUtil.readSegments(stream, segmentIdentifiers);
if (!photoshop.isEmpty()) {
  return readPathFromPhotoshopResources(new MemoryCacheImageInputStream(photoshop.get(0).data()));
}
```
So for this file `null` is returned because the path information is stored in another segment.

I've tried to work around this by processing all segments from the List but this fails with an exception in `com.twelvemonkeys.imageio.metadata.psd.PSDReader#read` in the switch statement.

I think it might be possible to join the data from all segments into one but haven't been successful in doing so.

**Comment by haraldk** [comment]
Hi Arno,

Thanks for reporting! And for providing a sample file, that makes it easy to reproduce the problem! 

And yes, I think you are spot on, that is indeed the problem. I wasn't aware that APP13/Photoshop 3.0 could be split in multiple segments like this, but it apparently can. 

The fix is rather easy, but we can't process each segment alone. They represent one, single Photoshop IRB, split into parts due to the JPEG segment size limitation (uses unsigned short for size). So we need to concatenate them and process the result.

The following works for me:

```java
            if (!photoshop.isEmpty()) {
                InputStream data = null;

                for (JPEGSegment ps : photoshop) {
                    data = data == null ? ps.data() : new SequenceInputStream(data, ps.data());
                }

                return readPathFromPhotoshopResources(new MemoryCacheImageInputStream(data));
            }
```

I'll create a fix for this as soon as possible, and try to make it into the next bug fix release. Can I use the file you provided for a test case?


Best regards,

-- 
Harald K

**Comment by aerpenbeck** [comment]
Thank's for your quick response. This looks promising, I'll try it out.

I have to clarify about whether I can contribute that particular file to your project.

**Comment by aerpenbeck** [comment]
There's a similar [issue](https://issues.apache.org/jira/browse/IMAGING-100) in the [Apache Commons Imaging](https://commons.apache.org/proper/commons-imaging/) project that has sample files with multiple APP13 segments attached, but those do not have a clipping path defined. Maybe those files could be useful for a test case though.

**Comment by haraldk** [comment]
Unfortunately, none of those images contains paths, so it won't help much for test cases for the clipping path code... Might use them for other tests though.

But the issue brings up another important factor, that multiple `APP13/Photoshop 3.0` segments may mean different things. Either a single split PSD IRB (as in your example), or actually multiple PSD IRBs (as your initial code suggestion would probably handle)... Then again, I don't think it matters much, as the PSD resources are just concatenated blocks anyway. 😀 


-- 
Harald K

**Comment by aerpenbeck** [comment]
Just to let you know that your example works for my test files, so I'd be happy to see it merged into the code base.

**Comment by haraldk** [comment]
The fix is part of the 3.6.3 release out now! 😀
